### PR TITLE
Add customisable conflict resolution for catalog 

### PR DESCRIPTION
**Pull request by sblausten**
### Background

Currently, conflicting entities ingested by a provider are dropped by the default processing engine. Conflicting entities are new additions which share the same name and namespace as another entity.

This logic can lead either to orphaned entities which have to be deleted manually or by a garbage collector, or to the entity disappearing completely until a refresh happens again (in the case of the events backend this relies on updates to the entity in the scm). 

Some real-world cases of this can be illustrated as follows (more detail here https://miro.com/app/board/uXjVPgNceZ4=/?share_link_id=803934569157 ):

1. **Renaming a catalog-info file**
    - Autodiscovery
      - With Garbage Collection - entity is removed and won't appear until next refresh at which point the correct state is regained
      - Without - entity is marked as orphan and won't receive updates
    - Events: Correctly moved
  - Cause: new entity is discarded as conflict, old entity is orphaned. 
  - Fixes: retry conflicts for period, overwrite on conflict
2. **Creating a duplicate entity in another file and then deleting the old file in two separate PRs**
    - Events and Autodiscovery:
      - With Garbage Collection - entity is removed and won't appear until next refresh at which point the correct state is regained
      - Without - entity is marked as orphan and won't receive updates
  - Cause: new entity is discarded as conflict, old entity is orphaned. 
  - Fixes: retry conflicts for period, overwrite on conflict + only delete on events if location key matches. 
3. **Rename an entity, then duplicate original in new file, in same PR or two PRs**
    - Events and Autodiscovery:
      - With Garbage Collection - only renamed entity is left, moved entity is removed and won't appear until next refresh at which point the correct state is regained
      - Without Garbage Collection  - moved entity is marked as orphan and won't receive updates
  - Cause: moved entity is discarded as conflict, old entity is orphaned. 
  - Fixes: retry conflicts for period, overwrite on conflict
4. **Switching two entities between two different files** 
    - Events and Autodiscovery:
      - With Garbage Collection - both entities are removed and won't appear until next refresh at which point the correct state is regained
      - Without Garbage Collection  - entities are marked as orphan and won't receive updates
  - Cause: refresh logic
  - Fixes: ???

### This PR

Passes a conflictHandler function through to the ProcessingDatabase.updateProcessedEntity interface, so that when conflicts are identified by trying to update an entity, the handler can implement custom behaviour, for instance retrying to ingest the conflict for a period, saving the conflict as an error on the original entity, or overwriting the existing entity with the conflicting update.

### Alternatives

- Instead of setting a custom handler, we could create a more opinionated approach where a settings based toggle could switch between predefined conflict resolution strategies.
- Sticking with the autodiscovery mechanism that offers eventual consistency if garbage collection is implemented but is not scaleable for larger organisations due to Github Rate limiting and also offers a worse user experience where updates can take a while to reconcile in the catalog, providing confusing intermediate states that may present a support/engineering burden on organisations running Backstage.

### Risks
Allowing user defined conflict handling pushes the burden of understanding the complexities of the ingestion and processing pipelines to implementers and provides more chance of mistakes and difficulties in implementing a sane approach.

### Example Resolution approaches
1. Schedule a task to retry ingesting the entity and use an AbortSignal to terminate the task after an hour or something. This probably needs some refactoring of the current code's api to work. 

2. Overwrite the entity when conflicting

```javascript
export default async function createPlugin(
  env: PluginEnvironment,
  providers?: Array<EntityProvider>,
): Promise<Router> {
  
  ...
  const conflictHandler = async (options: ConflictHandlerOptions) => {
    options.logger.info('Handling conflict with custom overwrite handler');

    const dbClient = await env.database.getClient();
    async function overwriteEntity(params: {
      db: Knex;
      entity: Entity;
      hash: string;
      originalLocationKey: string;
      newLocationKey: string;
    }): Promise<void> {
      const { db, entity, hash, originalLocationKey, newLocationKey } = params;

      const entityRef = stringifyEntityRef(entity);
      const serializedEntity = JSON.stringify(entity);
      const now = db.fn.now();

      await db('refresh_state')
        .update({
          unprocessed_entity: serializedEntity,
          unprocessed_hash: hash,
          location_key: newLocationKey,
          last_discovery_at: now,
          next_update_at: now,
        })
        .where('entity_ref', entityRef)
        .where('location_key', originalLocationKey);
    }

    return await overwriteEntity({
      db: dbClient,
      entity: options.entity,
      hash: options.hash,
      newLocationKey: options.newLocationKey,
      originalLocationKey: options.originalLocationKey,
    });
  };
  
  const { processingEngine, router } = await builder
    .setConflictHandler(conflictHandler)
    .build();

  await processingEngine.start();
  ...
```

NB: this would not resolve cases where two conflicting entities remained in the SCM under two files or in a single file. In this case the entities would flip flop on each autodiscovery loop or on update events for the entities. 
Deletion logic implemented by a custom garabage collector and/or future event based deletion will need to also make sure that the entity being deleted matches the location key of the current entity. This is because if the current entity has just been overriden by the update with a new file location for instance, the deletion should be invalid, otherwise we will face the same case as we currently do where we are left with no entity at all. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by jhaals** [comment]
Hi Sam!

Thanks for bringing this up. There's is definitely a gap in catalog when it comes to renaming and moving entities that needs addressing. There's is really no good way of doing it currently and you have situations where the entity is temporary removed before re-appearing. I think we need to map out and decide how to handle those intentional situations when a user wants to rename files or move a file or rename a repository as we want to keep some notion of conflicts as it could otherwise have security implications with entity takeovers. Straight up adding a conflictHandler to me feels like we're asking the users to do work that the catalog should be able to handle in the first place.

Would be great to get additional input from @backstage/catalog-core on this matter and if this is something that should be brought up in the Catalog SIG and/or drafted in an RFC fashion beforehand.

**Comment by sblausten** [comment]
Hi @jhaals 

Thanks for taking a look! 

> Straight up adding a conflictHandler to me feels like we're asking the users to do work that the catalog should be able to handle in the first place.
Agree on this. Injecting an optional conflictHandler seemed to me the least intrusive and lowest effort approach to start allowing users to handle cases that they are coming up against. I mentioned in the section on Alternatives:
> Instead of injecting the handler from the builder, we could create a more opinionated approach where a settings based toggle could switch between predefined conflict resolution strategies.
For now I'd think this would look something like using the injectable overwrite example by default for when event-based ingestion is enabled and a conflict is detected. I can look at making that change if you prefer the opinionated approach and we can discuss it next week. 

> if this is something that should be brought up in the Catalog SIG and/or drafted in an RFC fashion beforehand.
Yes definitely happy to talk this through at next weeks Catalog SIG. I followed the RFC format in the PR and have attempted to map out most of the intentional situations when a user wants to rename files, move a file, move an entity or rename a repository which seem to me to all have the same issue (conflict handling) and same fix for event based ingestion (overwriting). I simplified this in the PR background for ease of consumption as the suggested change was not opinionated, but I can elaborate on each of the individual cases.

> we want to keep some notion of conflicts as it could otherwise have security implications with entity takeovers
Yes for sure - I think users should be able to choose how they want to handle conflicts. I don't think there is any persisted notion of conflicts in the current backend state. I'm not currently proposing we stop handling conflicts in a specific way for all ingestion scenarios. But allowing customisable handling could allow for instance writing conflicts to an error in the `refresh_state` table that could be acted on by a separate processor, or just the overwrite strategy I'm proposing for event based ingestion.

File: plugins/catalog-backend/src/service/CatalogBuilder.ts [hunk 1]
```diff
@@ -420,7 +421,9 @@ export class CatalogBuilder {
   /**
    * Wires up and returns all of the component parts of the catalog
    */
-  async build(): Promise<{
+  async build(
+    conflictHandler?: (options: ConflictHandlerOptions) => Promise<void>,
```
Comment (sblausten) [inline plugins/catalog-backend/src/service/CatalogBuilder.ts:425]: An alternative approach could be to add this to the CatalogEnv. Not sure which is preferrable

Comment (freben) [reply plugins/catalog-backend/src/service/CatalogBuilder.ts:425]: The builder already has a large number of setters for different types of adaptations. None of them are on the build method. Any particular reason you put it there?

Comment (sblausten) [reply plugins/catalog-backend/src/service/CatalogBuilder.ts:425]: Ah good point yes thats much better

**Comment by freben** [comment]
For the sake of my understanding ... what code would _you_ put in the resolver? What is your need and your intended solution to that need? Sorry if you've already described it, in that case I missed it.

**Comment by sblausten** [comment]
> For the sake of my understanding ... what code would _you_ put in the resolver? What is your need and your intended solution to that need? Sorry if you've already described it, in that case I missed it.

Just the overwrite strategy given in the example and then make sure deletes only happen if the `location_key` of the entity being deleted matches that in the `refresh_state` table, as otherwise there can be a state where a delete intended as a clean-up removes the updated entity and you are left with nothing in the catalog.

**Comment by sabinasiddiqi** [comment]
Were there any updates on this discussion? We are also looking for.a way to resolve conflicts in a custom way but ideally dont want to have to fork this whole code to do that.

**Pull request closed without merging**